src/main/java/com/aws/greengrass/shadowmanager/sync/IotDataPlaneClientFactory.java

```diff
@@ -87,12 +72,20 @@ private String getIotCoreDataPlaneEndpoint(String iotDataEndpoint) {
         return String.format(IOT_CORE_DATA_PLANE_ENDPOINT_FORMAT, iotDataEndpoint);
     }
 
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException"})
+    private void configureClient() {
+        // If the crypto key provider service (such as pkcs11 provider component) is not available when the http client
+        // is created, there's a possibility that the client is not configured with mTLS.
 
-    private boolean validPath(Node node, String key) {
-        return validString(node, key) && Files.exists(Paths.get(key));
-    }
-
-    private void configureClient(DeviceConfiguration deviceConfiguration) {
+        // To ensure that the http client is configured with mutual auth, wait for the crypto key provider service
+        // to load. If the service is not loaded even after retrying, we continue with the client creation as shadow
+        // manager can still work locally.
+        try {
+            waitForCryptoKeyServiceProvider();
+        } catch (Exception e) {
+            logger.atError().log("Failed to get the key managers from crypto key provider service. Data plane"
```

Comment:
why are we returning an invalid client, ever?

Shadow manager will need to backoff and retry, waiting until the client is properly configured.

Reply:
Agree that this class should be retrying anything or blocking. We don't know what thread context we're running in. Blocking a thread is not appropriate.

From an interface perspective, this class should only return a client if it is actually usable. We'll need to think carefully about how to handle this at the upper layers

Reply:
Update the PR to throw a new exception called - `IoTDataPlaneClientCreation` when the service is not ready. All the cloud requests failing with this exception type are now retried.